File: shared/tyrum-shared/src/telegram.rs
use chrono::{DateTime, Utc};
use serde::Deserialize;
use serde::de::IgnoredAny;
use thiserror::Error;

use crate::{
    MediaKind, MessageContent, MessageSource, NormalizedMessage, NormalizedThread,
    NormalizedThreadMessage, PiiField, SenderMetadata, ThreadKind,
};

/// Normalize a raw Telegram update payload into the shared thread/message schema.
pub fn normalize_update(
    payload: &[u8],
) -> Result<NormalizedThreadMessage, TelegramNormalizationError> {
    let update: TelegramUpdate =
        serde_json::from_slice(payload).map_err(TelegramNormalizationError::InvalidPayload)?;
    normalize_parsed_update(update)
}

/// Normalize a pre-deserialized Telegram update.
pub fn normalize_parsed_update(
    update: TelegramUpdate,
) -> Result<NormalizedThreadMessage, TelegramNormalizationError> {
    let message = update
        .edited_message
        .or(update.message)
        .ok_or(TelegramNormalizationError::MissingMessage)?;

    let thread = to_normalized_thread(&message.chat);
    let timestamp = to_datetime(message.date)?;
    let edited_timestamp = match message.edit_date {
        Some(ts) => Some(to_datetime(ts)?),
        None => None,
    };

    let content = extract_content(&message)?;
    let mut message_pii = pii_from_content(&content);

    let sender = message.from.as_ref().map(|user| {
        if user.first_name.is_some() {
            message_pii.push(PiiField::SenderFirstName);
        }
        if user.last_name.is_some() {
            message_pii.push(PiiField::SenderLastName);
        }
        if user.username.is_some() {
            message_pii.push(PiiField::SenderUsername);
        }
        if user.language_code.is_some() {
            message_pii.push(PiiField::SenderLanguageCode);
        }

        SenderMetadata {
            id: user.id.to_string(),
            is_bot: user.is_bot,
            first_name: user.first_name.clone(),
            last_name: user.last_name.clone(),
            username: user.username.clone(),
            language_code: user.language_code.clone(),
        }
    });

    let normalized = NormalizedMessage {
        id: message.message_id.to_string(),
        thread_id: thread.id.clone(),
        source: MessageSource::Telegram,
        content,
        sender,
        timestamp,
        edited_timestamp,
        pii_fields: message_pii,
    };

    Ok(NormalizedThreadMessage {
        thread,
        message: normalized,
    })
}

/// Errors yielded when normalizing Telegram updates.
#[derive(Debug, Error)]
pub enum TelegramNormalizationError {
    #[error("failed to deserialize telegram update: {0}")]
    InvalidPayload(serde_json::Error),
    #[error("telegram update did not include a message or edited_message payload")]
    MissingMessage,
    #[error("unable to map telegram content into normalized schema")]
    UnsupportedContent,
    #[error("encountered invalid unix timestamp: {0}")]
    InvalidTimestamp(i64),
}

#[derive(Debug, Deserialize)]
pub struct TelegramUpdate {
    pub update_id: i64,
    pub message: Option<TelegramMessage>,
    pub edited_message: Option<TelegramMessage>,
}

#[derive(Debug, Deserialize)]
pub struct TelegramMessage {
    pub message_id: i64,
    pub date: i64,
    pub edit_date: Option<i64>,
    pub from: Option<TelegramUser>,
    pub chat: TelegramChat,
    pub text: Option<String>,
    pub caption: Option<String>,
    #[serde(default)]
    pub photo: Vec<IgnoredAny>,
    pub animation: Option<IgnoredAny>,
    pub audio: Option<IgnoredAny>,
    pub document: Option<IgnoredAny>,
    pub video: Option<IgnoredAny>,
    pub voice: Option<IgnoredAny>,
    pub video_note: Option<IgnoredAny>,
    pub sticker: Option<IgnoredAny>,
}

#[derive(Debug, Deserialize)]
pub struct TelegramUser {
    pub id: i64,
    #[serde(default)]
    pub is_bot: bool,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    pub username: Option<String>,
    pub language_code: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct TelegramChat {
    pub id: i64,
    #[serde(rename = "type")]
    pub kind: String,
    pub title: Option<String>,
    pub username: Option<String>,
}

fn to_normalized_thread(chat: &TelegramChat) -> NormalizedThread {
    let mut pii_fields = Vec::new();
    if chat.title.is_some() {
        pii_fields.push(PiiField::ThreadTitle);
    }
    if chat.username.is_some() {
        pii_fields.push(PiiField::ThreadUsername);
    }

    NormalizedThread {
        id: chat.id.to_string(),
        kind: match chat.kind.as_str() {
            "private" => ThreadKind::Private,
            "group" => ThreadKind::Group,
            "supergroup" => ThreadKind::Supergroup,
            "channel" => ThreadKind::Channel,
            _ => ThreadKind::Other,
        },
        title: chat.title.clone(),
        username: chat.username.clone(),
        pii_fields,
    }
}

fn extract_content(
    message: &TelegramMessage,
) -> Result<MessageContent, TelegramNormalizationError> {
    if let Some(text) = &message.text {
        return Ok(MessageContent::Text { text: text.clone() });
    }

    if let Some(media_kind) = infer_media_kind(message) {
        return Ok(MessageContent::MediaPlaceholder {
            media_kind,
            caption: message.caption.clone(),
        });
    }

    if let Some(caption) = &message.caption {
        return Ok(MessageContent::MediaPlaceholder {
            media_kind: MediaKind::Unknown,
            caption: Some(caption.clone()),
        });
    }

    Err(TelegramNormalizationError::UnsupportedContent)
}

fn infer_media_kind(message: &TelegramMessage) -> Option<MediaKind> {
    if !message.photo.is_empty() {
        return Some(MediaKind::Photo);
    }
    if message.video.is_some() {
        return Some(MediaKind::Video);
    }
    if message.animation.is_some() {
        return Some(MediaKind::Animation);
    }
    if message.document.is_some() {
        return Some(MediaKind::Document);
    }
    if message.audio.is_some() {
        return Some(MediaKind::Audio);
    }
    if message.voice.is_some() {
        return Some(MediaKind::Voice);
    }
    if message.video_note.is_some() {
        return Some(MediaKind::VideoNote);
    }
    if message.sticker.is_some() {
        return Some(MediaKind::Sticker);
    }

    None
}

fn pii_from_content(content: &MessageContent) -> Vec<PiiField> {
    match content {
        MessageContent::Text { .. } => vec![PiiField::MessageText],
        MessageContent::MediaPlaceholder { caption, .. } => {
            let mut fields = Vec::new();
            if caption.is_some() {
                fields.push(PiiField::MessageCaption);
            }
            fields
        }
    }
}

fn to_datetime(timestamp: i64) -> Result<DateTime<Utc>, TelegramNormalizationError> {
    DateTime::from_timestamp(timestamp, 0)
        .ok_or(TelegramNormalizationError::InvalidTimestamp(timestamp))
}

#[cfg(test)]
mod telegram_normalization {
    use super::*;
    use chrono::{TimeZone, Utc};
    use pretty_assertions::assert_eq;

    #[test]
    fn normalizes_text_message() {
        let payload = include_bytes!("../../tests/fixtures/telegram/text_message.json");
        let update = normalize_update(payload).expect("normalize text message");

        let expected_thread = NormalizedThread {
            id: "987654321".into(),
            kind: ThreadKind::Private,
            title: None,
            username: None,
            pii_fields: Vec::new(),
        };
        assert_eq!(update.thread, expected_thread);

        let expected_message = NormalizedMessage {
            id: "111".into(),
            thread_id: "987654321".into(),
            source: MessageSource::Telegram,
            content: MessageContent::Text {
                text: "Hello planner".into(),
            },
            sender: Some(SenderMetadata {
                id: "555555".into(),
                is_bot: false,
                first_name: Some("Ron".into()),
                last_name: Some("Swanson".into()),
                username: Some("rons".into()),
                language_code: Some("en".into()),
            }),
            timestamp: Utc.timestamp_opt(1_710_000_000, 0).single().unwrap(),
            edited_timestamp: None,
            pii_fields: vec![
                PiiField::MessageText,
                PiiField::SenderFirstName,
                PiiField::SenderLastName,
                PiiField::SenderUsername,
                PiiField::SenderLanguageCode,
            ],
        };

        assert_eq!(update.message, expected_message);
    }

    #[test]
    fn normalizes_edited_message() {
        let payload = include_bytes!("../../tests/fixtures/telegram/edited_message.json");
        let update = normalize_update(payload).expect("normalize edited message");

        assert_eq!(
            update.message.edited_timestamp,
            Some(Utc.timestamp_opt(1_710_000_600, 0).single().unwrap())
        );
        assert!(matches!(
            update.message.content,
            MessageContent::Text { ref text } if text == "Hello planner edited"
        ));
        assert!(update.message.pii_fields.contains(&PiiField::MessageText));
    }

    #[test]
    fn normalizes_media_message() {
        let payload = include_bytes!("../../tests/fixtures/telegram/media_message.json");
        let update = normalize_update(payload).expect("normalize media message");

        assert_eq!(update.thread.kind, ThreadKind::Supergroup);
        assert!(update.thread.pii_fields.contains(&PiiField::ThreadTitle));

        match update.message.content {
            MessageContent::MediaPlaceholder {
                media_kind,
                caption,
            } => {
                assert_eq!(media_kind, MediaKind::Photo);
                assert_eq!(caption.as_deref(), Some("Check this out"));
            }
            other => panic!("unexpected content: {other:?}"),
        }

        assert!(
            update
                .message
                .pii_fields
                .contains(&PiiField::MessageCaption)
        );
    }

    #[test]
    fn normalizes_unknown_media_with_caption() {
        let payload = include_bytes!("../../tests/fixtures/telegram/unknown_media_caption.json");
        let update = normalize_update(payload).expect("normalize unknown media");

        match update.message.content {
            MessageContent::MediaPlaceholder {
                media_kind,
                caption,
            } => {
                assert_eq!(media_kind, MediaKind::Unknown);
                assert_eq!(caption.as_deref(), Some("Future media caption"));
            }
            other => panic!("unexpected content: {other:?}"),
        }

        assert!(
            update
                .message
                .pii_fields
                .contains(&PiiField::MessageCaption)
        );
    }

    #[test]
    fn rejects_unknown_payload() {
        let payload = br#"{"update_id": 1}"#;
        let err = normalize_update(payload).expect_err("unsupported payload");
        assert!(matches!(err, TelegramNormalizationError::MissingMessage));
    }
}

File: packages/gateway/src/modules/ingress/telegram.ts
/**
 * Telegram ingress normalizer.
 *
 * Converts raw Telegram update payloads into the NormalizedThreadMessage schema.
 */

import { normalizedContainerKindFromThreadKind } from "@tyrum/contracts";
import type {
  MediaKind,
  MessageContent,
  NormalizedAttachment,
  NormalizedMessage,
  NormalizedMessageEnvelope,
  NormalizedThread,
  NormalizedThreadMessage,
  PiiField,
  SenderMetadata,
  ThreadKind,
} from "@tyrum/contracts";
import { telegramAccountIdFromEnv } from "../channels/telegram-account.js";
import type { ArtifactStore } from "../artifact/store.js";
import type { TelegramBot } from "./telegram-bot.js";

export class TelegramNormalizationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TelegramNormalizationError";
  }
}

interface TelegramUpdate {
  update_id: number;
  message?: TelegramMessage;
  edited_message?: TelegramMessage;
}

interface TelegramMessage {
  message_id: number;
  date: number;
  edit_date?: number;
  from?: TelegramUser;
  chat: TelegramChat;
  text?: string;
  caption?: string;
  photo?: TelegramPhotoSize[];
  animation?: TelegramFileLike;
  audio?: TelegramFileLike;
  document?: TelegramFileLike;
  video?: TelegramFileLike;
  voice?: TelegramFileLike;
  video_note?: TelegramFileLike;
  sticker?: TelegramFileLike;
}

interface TelegramPhotoSize {
  file_id: string;
  file_unique_id?: string;
  file_size?: number;
  width?: number;
  height?: number;
}

interface TelegramFileLike {
  file_id: string;
  file_unique_id?: string;
  file_name?: string;
  mime_type?: string;
  file_size?: number;
  width?: number;
  height?: number;
  duration?: number;
}

interface TelegramUser {
  id: number;
  is_bot: boolean;
  first_name?: string;
  last_name?: string;
  username?: string;
  language_code?: string;
}

interface TelegramChat {
  id: number;
  type: string;
  title?: string;
  username?: string;
}

type TelegramMediaCandidate = {
  kind: MediaKind;
  fileId: string;
  fileUniqueId?: string;
  filename?: string;
  mimeType?: string;
  sizeBytes?: number;
  width?: number;
  height?: number;
  duration?: number;
};

export type TelegramMediaNormalizationDeps = {
  telegramBot: TelegramBot;
  artifactStore: ArtifactStore;
  maxUploadBytes?: number;
};

function mapChatType(type: string): ThreadKind {
  switch (type) {
    case "private":
      return "private";
    case "group":
      return "group";
    case "supergroup":
      return "supergroup";
    case "channel":
      return "channel";
    default:
      return "other";
  }
}

function toDatetime(timestamp: number): string {
  const date = new Date(timestamp * 1000);
  if (Number.isNaN(date.getTime())) {
    throw new TelegramNormalizationError(`encountered invalid unix timestamp: ${timestamp}`);
  }
  return date.toISOString();
}

function toNormalizedThread(chat: TelegramChat): NormalizedThread {
  const piiFields: PiiField[] = [];
  if (chat.title != null) {
    piiFields.push("thread_title");
  }
  if (chat.username != null) {
    piiFields.push("thread_username");
  }

  return {
    id: String(chat.id),
    kind: mapChatType(chat.type),
    title: chat.title,
    username: chat.username,
    pii_fields: piiFields,
  };
}

function trimNonEmpty(value: string | undefined): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function deserializeUpdate(payload: string | Uint8Array): TelegramUpdate {
  try {
    const raw = typeof payload === "string" ? payload : new TextDecoder().decode(payload);
    return JSON.parse(raw) as TelegramUpdate;
  } catch (err) {
    throw new TelegramNormalizationError(
      `failed to deserialize telegram update: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

function selectMessage(update: TelegramUpdate): TelegramMessage {
  const message = update.edited_message ?? update.message;
  if (message == null) {
    throw new TelegramNormalizationError(
      "telegram update did not include a message or edited_message payload",
    );
  }
  return message;
}

function inferMediaCandidate(message: TelegramMessage): TelegramMediaCandidate | undefined {
  const photo = message.photo?.at(-1);
  if (photo?.file_id) {
    return {
      kind: "photo",
      fileId: photo.file_id,
      fileUniqueId: photo.file_unique_id,
      mimeType: "image/jpeg",
      sizeBytes: photo.file_size,
      width: photo.width,
      height: photo.height,
    };
  }

  const pickFile = (
    kind: MediaKind,
    file: TelegramFileLike | undefined,
    fallbackMimeType?: string,
  ): TelegramMediaCandidate | undefined => {
    if (!file?.file_id) {
      return undefined;
    }
    return {
      kind,
      fileId: file.file_id,
      fileUniqueId: file.file_unique_id,
      filename: file.file_name,
      mimeType: file.mime_type ?? fallbackMimeType,
      sizeBytes: file.file_size,
      width: file.width,
      height: file.height,
      duration: file.duration,
    };
  };

  return (
    pickFile("video", message.video) ??
    pickFile("animation", message.animation, "video/mp4") ??
    pickFile("document", message.document) ??
    pickFile("audio", message.audio) ??
    pickFile("voice", message.voice, "audio/ogg") ??
    pickFile("video_note", message.video_note, "video/mp4") ??
    pickFile("sticker", message.sticker, "image/webp")
  );
}

function toContent(text: string | undefined, attachments: NormalizedAttachment[]): MessageContent {
  if (typeof text !== "string" && attachments.length === 0) {
    throw new TelegramNormalizationError(
      "telegram update did not include text or a recognized attachment",
    );
  }
  return {
    ...(text ? { text } : {}),
    attachments,
  };
}

function piiFromContent(content: MessageContent): PiiField[] {
  return typeof content.text === "string" ? ["message_text"] : [];
}

function toEnvelopeContent(
  content: MessageContent,
): NormalizedMessageEnvelope["content"] | undefined {
  if (typeof content.text !== "string" && content.attachments.length === 0) {
    return undefined;
  }
  return content;
}

async function materializeTelegramAttachment(
  deps: TelegramMediaNormalizationDeps,
  message: TelegramMessage,
  candidate: TelegramMediaCandidate,
): Promise<NormalizedAttachment> {
  if (
    typeof deps.maxUploadBytes === "number" &&
    typeof candidate.sizeBytes === "number" &&
    candidate.sizeBytes > deps.maxUploadBytes
  ) {
    throw new TelegramNormalizationError(
      `telegram attachment exceeds maxUploadBytes (${String(candidate.sizeBytes)} > ${String(deps.maxUploadBytes)})`,
    );
  }
  const downloaded = await deps.telegramBot.downloadFileById(candidate.fileId);
  if (typeof deps.maxUploadBytes === "number" && downloaded.body.byteLength > deps.maxUploadBytes) {
    throw new TelegramNormalizationError(
      `telegram attachment exceeds maxUploadBytes (${String(downloaded.body.byteLength)} > ${String(deps.maxUploadBytes)})`,
    );
  }
  const artifact = await deps.artifactStore.put({
    kind: "file",
    body: downloaded.body,
    mime_type: candidate.mimeType ?? downloaded.mediaType,
    filename: candidate.filename,
    metadata: {
      source: "telegram-ingress",
      telegram: {
        channel_kind: candidate.kind,
        chat_id: String(message.chat.id),
        file_id: candidate.fileId,
        file_unique_id: candidate.fileUniqueId,
        height: candidate.height,
        message_id: String(message.message_id),
        size_bytes: candidate.sizeBytes,
        width: candidate.width,
        duration: candidate.duration,
      },
    },
  });
  return {
    ...artifact,
    channel_kind: candidate.kind,
  };
}

function buildNormalizedMessage(input: {
  message: TelegramMessage;
  thread: NormalizedThread;
  sender?: SenderMetadata;
  timestamp: string;
  editedTimestamp?: string;
  content: MessageContent;
  messagePii: PiiField[];
}): NormalizedMessage {
  return {
    id: String(input.message.message_id),
    thread_id: input.thread.id,
    source: "telegram",
    content: input.content,
    sender: input.sender,
    timestamp: input.timestamp,
    edited_timestamp: input.editedTimestamp,
    pii_fields: input.messagePii,
    envelope: (() => {
      const envelopeContent = toEnvelopeContent(input.content);
      if (!envelopeContent) {
        return undefined;
      }

      return {
        message_id: String(input.message.message_id),
        received_at: input.timestamp,
        delivery: {
          channel: "telegram",
          account: telegramAccountIdFromEnv(),
        },
        container: {
          kind: normalizedContainerKindFromThreadKind(input.thread.kind),
          id: input.thread.id,
        },
        sender: {
          id: input.sender?.id ?? `chat:${input.thread.id}`,
          ...(input.sender?.username != null ? { display: input.sender.username } : {}),
        },
        content: envelopeContent,
        provenance: ["user"],
      };
    })(),
  };
}

function buildSender(input: {
  message: TelegramMessage;
  messagePii: PiiField[];
}): SenderMetadata | undefined {
  if (!input.message.from) {
    return undefined;
  }

  const user = input.message.from;
  if (user.first_name != null) {
    input.messagePii.push("sender_first_name");
  }
  if (user.last_name != null) {
    input.messagePii.push("sender_last_name");
  }
  if (user.username != null) {
    input.messagePii.push("sender_username");
  }
  if (user.language_code != null) {
    input.messagePii.push("sender_language_code");
  }

  return {
    id: String(user.id),
    is_bot: user.is_bot,
    first_name: user.first_name,
    last_name: user.last_name,
    username: user.username,
    language_code: user.language_code,
  };
}

function normalizeMessageCore(payload: string | Uint8Array): {
  message: TelegramMessage;
  thread: NormalizedThread;
  timestamp: string;
  editedTimestamp?: string;
} {
  const update = deserializeUpdate(payload);
  const message = selectMessage(update);
  return {
    message,
    thread: toNormalizedThread(message.chat),
    timestamp: toDatetime(message.date),
    editedTimestamp: message.edit_date != null ? toDatetime(message.edit_date) : undefined,
  };
}

export function normalizeUpdate(payload: string | Uint8Array): NormalizedThreadMessage {
  const { message, thread, timestamp, editedTimestamp } = normalizeMessageCore(payload);
  const text = trimNonEmpty(message.text ?? message.caption);
  const content = toContent(text, []);
  const messagePii = piiFromContent(content);
  const sender = buildSender({ message, messagePii });

  return {
    thread,
    message: buildNormalizedMessage({
      message,
      thread,
      sender,
      timestamp,
      editedTimestamp,
      content,
      messagePii,
    }),
  };
}

export async function normalizeUpdateWithMedia(
  payload: string | Uint8Array,
  deps: TelegramMediaNormalizationDeps,
): Promise<NormalizedThreadMessage> {
  const { message, thread, timestamp, editedTimestamp } = normalizeMessageCore(payload);
  const text = trimNonEmpty(message.text ?? message.caption);
  const attachments: NormalizedAttachment[] = [];
  const candidate = inferMediaCandidate(message);
  if (candidate) {
    attachments.push(await materializeTelegramAttachment(deps, message, candidate));
  }

  const content = toContent(text, attachments);
  const messagePii = piiFromContent(content);
  const sender = buildSender({ message, messagePii });

  return {
    thread,
    message: buildNormalizedMessage({
      message,
      thread,
      sender,
      timestamp,
      editedTimestamp,
      content,
      messagePii,
    }),
  };
}
